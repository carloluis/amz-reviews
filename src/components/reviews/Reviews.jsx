import React from 'react';
import PropTypes from 'prop-types';
import Button from './button/Button';
import Review from './review/Review';
import Loading from './loading/Loading';
import styles from './Reviews.scss';

class Reviews extends React.Component {
    constructor(props) {
        super(props);
        this.handleScroll = this.handleScroll.bind(this);
    }
    componentDidMount() {
        this.props.refreshReviews();
        window.addEventListener('scroll', this.handleScroll);
    }
    componentWillUnmount() {
        window.removeEventListener('scroll', this.handleScroll);
    }
    handleScroll() {
        const { loading, page, fetchReviews, hasMore } = this.props;
        if (!hasMore) {
            window.removeEventListener('scroll', this.handleScroll);
        }

        const { documentElement } = document;
        if (!loading && documentElement.scrollHeight - documentElement.scrollTop === documentElement.clientHeight) {
            fetchReviews(page + 1);
        }
    }
    render() {
        const { loading, reviews, refreshReviews } = this.props;

        const reviewsElements = reviews.map(({ key, values }) => (
            <React.Fragment key={key}>
                {!!key && <div className={styles.groupInfo}>{key}</div>}
                {values.map(review => <Review key={review.reviewId} {...review} />)}
            </React.Fragment>
        ));

        return (
            <section className={styles.container}>
                <Button text="REFRESH" onClick={refreshReviews} />
                <div>{reviewsElements}</div>
                <div className={styles.loading}>
                    <Loading loading={loading} />
                </div>
            </section>
        );
    }
}

Reviews.propTypes = {
    refreshReviews: PropTypes.func.isRequired,
    fetchReviews: PropTypes.func.isRequired,
    reviews: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    hasMore: PropTypes.bool.isRequired,
    page: PropTypes.number.isRequired,
    error: PropTypes.bool.isRequired
};

export default Reviews;
