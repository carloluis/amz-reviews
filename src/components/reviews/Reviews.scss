@import '../../styles/palette';

.container {
    margin: 32px auto;

    .loading {
        color: $dark-font;
        text-align: center;
    }
}
