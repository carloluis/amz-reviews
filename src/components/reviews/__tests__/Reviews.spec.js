import React from 'react';
import renderer from 'react-test-renderer';
import Reviews from '../Reviews';

const PROPS = {
    refreshReviews: jest.fn(),
    fetchReviews: jest.fn(),
    reviews: [],
    loading: false,
    hasMore: false,
    page: 0,
    error: false
};

describe('<Reviews/>', () => {
    beforeEach(() => {
        PROPS.refreshReviews.mockClear();
        PROPS.fetchReviews.mockClear();
    });
    afterEach(() => {
        expect(PROPS.refreshReviews).toHaveBeenCalled();
    });
    it('snapshot', () => {
        const result = renderer.create(<Reviews {...PROPS} />);
        expect(result).toMatchSnapshot();
    });
    it('snapshot - loading', () => {
        const result = renderer.create(<Reviews {...PROPS} loading />);
        expect(result).toMatchSnapshot();
    });
    it('snapshot - with reviews', () => {
        const reviews = [
            {
                key: '',
                values: [
                    {
                        childAsin: 'childAsin?',
                        content: 'Lorem ipsum',
                        created: 1528940971610,
                        productTitle: 'Product Title',
                        productImg: 'Image',
                        reviewId: 'Review Id',
                        reviewCreated: 1528940971610,
                        stars: 5,
                        title: 'Title'
                    }
                ]
            }
        ];
        const result = renderer.create(<Reviews {...PROPS} reviews={reviews} page={1} hasMore />);
        expect(result).toMatchSnapshot();
    });
});
