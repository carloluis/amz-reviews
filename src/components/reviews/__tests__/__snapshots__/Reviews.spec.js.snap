// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews/> snapshot - loading 1`] = `
<section
  className="container"
>
  <a
    className="button"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    REFRESH
  </a>
  <div />
  <div
    className="loading"
  >
    <div
      className="dots"
    />
  </div>
</section>
`;

exports[`<Reviews/> snapshot - with reviews 1`] = `
<section
  className="container"
>
  <a
    className="button"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    REFRESH
  </a>
  <div>
    <div
      className="container"
    >
      <div
        className="header"
      >
        <i
          className="image"
        />
        <div
          className="container"
        >
          <div
            className="title"
          >
            DATE
          </div>
          <div
            className="text"
          >
            13.6.2018
          </div>
        </div>
        <div
          className="container"
        >
          <div
            className="title"
          >
            STARS
          </div>
          <div
            className="text"
          >
            ★★★★★
          </div>
        </div>
        <div
          className="container"
        >
          <div
            className="title"
          >
            childAsin?
          </div>
          <div
            className="text"
          >
            Product Title
          </div>
        </div>
      </div>
      <div
        className="content"
      >
        <span>
          Title
        </span>
        <p>
          Lorem ipsum
        </p>
      </div>
    </div>
  </div>
  <div
    className="loading"
  />
</section>
`;

exports[`<Reviews/> snapshot 1`] = `
<section
  className="container"
>
  <a
    className="button"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    REFRESH
  </a>
  <div />
  <div
    className="loading"
  />
</section>
`;
